File: tests/sysid32/Reject.cs
using System;
using System.IO;
class Reject {
    static void Main(string[] args) {
        foreach (string path in Directory.GetFiles(args[0], "*.v2")) {
            byte[] original = File.ReadAllBytes(path);
            int rejectedLength = File.ReadAllBytes(Path.ChangeExtension(path, "frame")).Length;
            foreach (bool magicTrailer in new[] { false, true }) {
                byte[] bytes = (byte[])original.Clone();
                // An early discard must not treat the final CRC/signature byte as STX.
                if (magicTrailer) bytes[rejectedLength - 1] = 0xfd;
                var stream = new MemoryStream(bytes);
                var parser = new MAVLink.MavlinkParse();
                if (parser.ReadPacket(stream) != null) throw new Exception("accepted " + path);
                if (parser.badCRC != 0 || parser.badIncompatFlags != 1)
                    throw new Exception("unsupported frame reported as corrupt " + path);
                if (stream.Position != rejectedLength)
                    throw new Exception("wrong discard length " + path + " magicTrailer=" + magicTrailer);
                var packet = parser.ReadPacket(stream);
                if (packet == null || packet.msgid != 0 || packet.sysid != 42 || stream.Position != stream.Length)
                    throw new Exception("lost synchronization " + path + " magicTrailer=" + magicTrailer);
            }
            // Packet-oriented transports must reject the same header too.
            bool rejected = false;
            try { new MAVLink.MAVLinkMessage(File.ReadAllBytes(Path.ChangeExtension(path, "frame"))); }
            catch (NotSupportedException) { rejected = true; }
            if (!rejected) throw new Exception("constructor accepted " + path);
        }
        byte[] corrupt = File.ReadAllBytes(Path.Combine(args[0], "legacy"));
        corrupt[corrupt.Length - 1] ^= 1;
        var crcParser = new MAVLink.MavlinkParse();
        if (crcParser.ReadPacket(new MemoryStream(corrupt)) != null ||
            crcParser.badCRC != 1 || crcParser.badIncompatFlags != 0)
            throw new Exception("CRC error misclassified");
    }
}

File: tests/sysid32/Reject.java
import com.MAVLink.Parser;
import com.MAVLink.MAVLinkPacket;
import java.nio.file.*;
public class Reject {
    public static void main(String[] args) throws Exception {
        try (DirectoryStream<Path> files = Files.newDirectoryStream(Paths.get(args[0]), "*.v2")) {
            for (Path file : files) {
                byte[] original = Files.readAllBytes(file);
                String frameName = file.getFileName().toString().replaceFirst("\\.v2$", ".frame");
                int rejectedLength = Files.readAllBytes(file.resolveSibling(frameName)).length;
                for (boolean magicTrailer : new boolean[]{false, true}) {
                    byte[] bytes = original.clone();
                    // A discard ending one byte early must not interpret the
                    // rejected frame's final CRC/signature byte as a new STX.
                    if (magicTrailer) bytes[rejectedLength - 1] = (byte)0xfd;
                    Parser parser = new Parser();
                    int count = 0;
                    for (byte b : bytes) {
                        MAVLinkPacket packet = parser.mavlink_parse_char(b & 255);
                        if (packet != null) {
                            if (packet.msgid != 0 || packet.sysid != 42) throw new AssertionError(file);
                            count++;
                        }
                    }
                    if (count != 1) throw new AssertionError(file + ": " + count + " magicTrailer=" + magicTrailer);
                }
            }
        }
    }
}
